### DISCUSSION TITLE
How to get `mpris:artUrl` from Mpris to send notifications on track change(Mpris2 Widget)

### DISCUSSION THREAD

**Discussion by Tanish2002** [qtile/qtile#4330] [Q&A]
The title is pretty bad. Yes. 

I would like to send track change notifications with an icon(album art/video thumbnail). Like Spotify does through the Mpris2 widget.
I was able to come up with this
```py
class MyMpris(widget.Mpris2):
    def parse_message(
        self,
        _interface_name: str,
        changed_properties: dict[str, Variant],
        _invalidated_properties: list[str],
    ):
        widget.Mpris2.parse_message(
            self,
            _interface_name,
            changed_properties,
            _invalidated_properties,
        )

        if self.track_info != self.text:
            send_notification("Now Playing", self.track_info)
```
This works fine. It sends notification with the track on change (Also weirdly it sends the same notification twice. IDK if it's a mpris issue or not). 

I would like to modify this widget to get `mpris:artUrl`
Example from playerctl
```
playerctl metadata mpris:artUrl
https://i1.ytimg.com/vi/PcbrKsAcwn4/hqdefault.jpg
```
I was thinking of fetching this image and sending a notification with this. Though Idk how I can modify the widget to get this artURL

Also slightly out of topic but the `send_notification` in utils should have a icon parameter as well.

**Comment by Tanish2002** [comment]
Update:
well this is flawed as it will send notification whenever I pause as well. 
```py
        if self.track_info != self.text:
            send_notification("Now Playing", self.track_info)
```

**Reply by elParaguayo** [reply]
The widget reacts to metadata changes and playback status changes so you'd need to filter out the latter if you don't want them.

**Reply by Tanish2002** [reply]
> The widget reacts to metadata changes and playback status changes so you'd need to filter out the latter if you don't want them.

How can I get the playback status changes?
Also now that I'm thinking if I can subscribe to to playback status changes then I would like to send notifications on pause and play as well.



Also my main problem currently is how I can get `mpris:artUrl`

**Reply by elParaguayo** [reply]
Look at the `parse_message` method in the widget.

The artwork will be an entry in the metadata dictionary.

**Reply by elParaguayo** [reply]
Also, qtile-extras has a version of the widget with a popup that displays artwork.

**Comment by Tanish2002** [comment]
> Look at the `parse_message` method in the widget.
> 
> The artwork will be an entry in the metadata dictionary.

Thanks @elParaguayo, I was able to hack together it. Leaving it here in case anyone needs it in the future:

Though I still get the notification thrown twice :thinking:. Would that be mpris issue? or something wrong in my implementation?
I'm just checking for `if self.track_info != self.text:`
```py
from libqtile import widget
import urllib.request
import notify2

class MyMpris(widget.Mpris2):
    def fetch_image(self, url):
        try:
            urllib.request.urlretrieve(url, "/tmp/playing.png")
            return "/tmp/playing.png"
        except:
            return ""

    def __init__(self, **config):
        widget.Mpris2.__init__(self, **config)
        notify2.init("qtile")

    def parse_message(
        self,
        _interface_name,
        changed_properties,
        _invalidated_properties,
    ):
        widget.Mpris2.parse_message(
            self,
            _interface_name,
            changed_properties,
            _invalidated_properties,
        )

        if self.track_info != self.text:
            playbackstatus = getattr(
                changed_properties.get("PlaybackStatus"), "value", None
            )
            # The player was just paused
            if playbackstatus:
                return

            metadata = changed_properties.get("Metadata")
            artwork = metadata.value.get("mpris:artUrl").value
            icon = ""
            if artwork:
                icon = self.fetch_image(artwork)

            notify2.Notification("Now Playing", self.track_info, icon).show()

            # Doesn't have a parameter for icon :(
            # send_notification("Now Playing", self.track_info)
```
